File: apps/ios/Packages/HushOSKit/Sources/HushOSKit/Tags.swift
import Foundation

/*
 * Tags: flat sets an item belongs to across folders. The workspace keeps one
 * registry as a sealed document under the workspace key (tags, and which
 * items carry each), so a share's recipient learns nothing of them and
 * tagging never touches a node's envelope.
 */
public struct Tag: Codable, Sendable, Identifiable, Hashable {
    public var id: String
    public var name: String
    public var colour: String
    public init(id: String, name: String, colour: String) {
        self.id = id
        self.name = name
        self.colour = colour
    }
}

public struct TagRegistry: Codable, Sendable, Equatable {
    public var version: Int
    public var tags: [Tag]
    /* Tag id -> the ids of the nodes carrying it. */
    public var items: [String: [String]]

    public static let empty = TagRegistry(version: 2, tags: [], items: [:])
    public static let presets = ["blue", "ink", "yellow", "teal", "coral"]
    public static let maxTags = 200
    public static let maxPerNode = 16

    public func tags(of nodeId: String) -> [Tag] {
        tags.filter { items[$0.id]?.contains(nodeId) ?? false }
    }

    public func nodes(with tagId: String) -> [String] { items[tagId] ?? [] }

    public mutating func add(name: String) throws -> Tag {
        let clean = try TagRegistry.checkName(name)
        if let existing = tags.first(where: { $0.name.caseInsensitiveCompare(clean) == .orderedSame }) { return existing }
        guard tags.count < TagRegistry.maxTags else { throw AuthError.message("A workspace can have at most 200 tags.") }
        let tag = Tag(id: UUID().uuidString.lowercased(), name: clean, colour: TagRegistry.presets[tags.count % TagRegistry.presets.count])
        tags.append(tag)
        return tag
    }

    public mutating func rename(_ id: String, to name: String) throws {
        let clean = try TagRegistry.checkName(name)
        guard let index = tags.firstIndex(where: { $0.id == id }) else { return }
        tags[index].name = clean
    }

    public mutating func recolour(_ id: String, to colour: String) {
        guard let index = tags.firstIndex(where: { $0.id == id }) else { return }
        tags[index].colour = colour
    }

    public mutating func remove(_ id: String) {
        tags.removeAll { $0.id == id }
        items[id] = nil
    }

    public mutating func assign(_ nodeId: String, tagIds: [String]) throws {
        guard tagIds.count <= TagRegistry.maxPerNode else { throw AuthError.message("An item can carry at most 16 tags.") }
        for tag in tags {
            var nodes = items[tag.id] ?? []
            nodes.removeAll { $0 == nodeId }
            if tagIds.contains(tag.id) { nodes.append(nodeId) }
            if nodes.isEmpty { items[tag.id] = nil } else { items[tag.id] = nodes }
        }
    }

    static func checkName(_ name: String) throws -> String {
        let clean = name.trimmingCharacters(in: .whitespaces).split(separator: " ").joined(separator: " ")
        let points = clean.unicodeScalars.count
        guard points >= 1 else { throw AuthError.message("Enter a tag name.") }
        guard points <= 40 else { throw AuthError.message("Use a tag name of at most 40 characters.") }
        return clean
    }
}

public struct DocumentView: Codable, Sendable {
    public let kind: String
    public let version: Int
    public let envelope: String
    public let updatedAt: String
}

extension DriveAPI {
    public func document(workspaceId: String, kind: String) async throws -> DocumentView? {
        struct Body: Decodable { let document: DocumentView? }
        return try await performPublic(try requestPublic("/workspaces/\(workspaceId)/documents/\(kind)"), as: Body.self).document
    }

    /* `version` is the one this write is based on; the server refuses a stale base with `stale`. */
    public func putDocument(workspaceId: String, kind: String, version: Int, envelope: String) async throws -> DocumentView {
        struct Body: Decodable { let document: DocumentView }
        return try await performPublic(try mutationPublic("/workspaces/\(workspaceId)/documents/\(kind)", method: "PUT", body: ["version": version, "envelope": envelope]), as: Body.self).document
    }
}

extension Vault {
    private static let tagsKind = "tags"

    /* The registry as the workspace holds it, and the version to base the next write on. */
    public func tags() async throws -> (registry: TagRegistry, version: Int) {
        let workspace = try await loadWorkspace()
        guard let key = workspaceKeyData() else { throw DriveAPIError.server(500, "Workspace not opened") }
        guard let document = try await api.document(workspaceId: workspace.workspaceId, kind: Self.tagsKind) else {
            return (.empty, 0)
        }
        let json = try documentOpen(
            ctx: DocumentContext(workspaceId: workspace.workspaceId, kind: Self.tagsKind, version: UInt64(document.version)),
            workspaceKey: key, envelope: try base64urlDecode(value: document.envelope)
        )
        let registry = try JSONDecoder().decode(TagRegistry.self, from: Data(json.utf8))
        return (registry, document.version)
    }

    /* Seals and writes the registry one version up; a stale base is reloaded and the change reapplied once. */
    public func saveTags(_ change: @Sendable (inout TagRegistry) throws -> Void) async throws -> TagRegistry {
        let workspace = try await loadWorkspace()
        guard let key = workspaceKeyData() else { throw DriveAPIError.server(500, "Workspace not opened") }
        var attempt = 0
        while true {
            var (registry, version) = try await tags()
            try change(&registry)
            let json = String(decoding: try JSONEncoder().encode(registry), as: UTF8.self)
            let envelope = try documentSeal(
                ctx: DocumentContext(workspaceId: workspace.workspaceId, kind: Self.tagsKind, version: UInt64(version + 1)),
                workspaceKey: key, json: json
            )
            do {
                _ = try await api.putDocument(workspaceId: workspace.workspaceId, kind: Self.tagsKind, version: version, envelope: base64urlEncode(bytes: envelope))
                return registry
            } catch DriveAPIError.server(409, _) where attempt == 0 {
                attempt += 1
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/com/hushos/app/data/Tags.kt
package com.hushos.app.data

import com.hushos.core.DocumentContext
import com.hushos.core.base64urlDecode
import com.hushos.core.base64urlEncode
import com.hushos.core.documentOpen
import com.hushos.core.documentSeal
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/*
 * Tags: flat sets an item belongs to across folders. The workspace keeps one
 * registry as a sealed document under the workspace key (tags, and which
 * items carry each), so a share's recipient learns nothing of them.
 */
data class Tag(val id: String, val name: String, val colour: String)

class TagRegistry(val tags: MutableList<Tag>, val items: MutableMap<String, MutableList<String>>) {
    companion object {
        val PRESETS = listOf("blue", "ink", "yellow", "teal", "coral")
        fun empty() = TagRegistry(mutableListOf(), mutableMapOf())

        fun parse(json: String): TagRegistry {
            val root = JSONObject(json)
            val tags = root.getJSONArray("tags").let { a -> (0 until a.length()).map { a.getJSONObject(it) }.map { Tag(it.getString("id"), it.getString("name"), it.getString("colour")) } }
            val items = mutableMapOf<String, MutableList<String>>()
            val raw = root.getJSONObject("items")
            for (key in raw.keys()) items[key] = raw.getJSONArray(key).let { a -> (0 until a.length()).map { a.getString(it) } }.toMutableList()
            return TagRegistry(tags.toMutableList(), items)
        }
    }

    fun toJson(): String {
        val items = JSONObject()
        for ((k, v) in this.items) items.put(k, JSONArray(v))
        return JSONObject().put("version", 2).put("tags", JSONArray(tags.map { JSONObject().put("id", it.id).put("name", it.name).put("colour", it.colour) })).put("items", items).toString()
    }

    fun tagsOf(nodeId: String): List<Tag> = tags.filter { items[it.id]?.contains(nodeId) == true }
    fun nodesWith(tagId: String): List<String> = items[tagId].orEmpty()

    fun add(name: String): Tag {
        val clean = checkName(name)
        tags.firstOrNull { it.name.equals(clean, ignoreCase = true) }?.let { return it }
        if (tags.size >= 200) throw AuthFailure("A workspace can have at most 200 tags.")
        val tag = Tag(UUID.randomUUID().toString(), clean, PRESETS[tags.size % PRESETS.size])
        tags.add(tag)
        return tag
    }

    fun rename(id: String, name: String) {
        val clean = checkName(name)
        val index = tags.indexOfFirst { it.id == id }
        if (index >= 0) tags[index] = tags[index].copy(name = clean)
    }

    fun recolour(id: String, colour: String) {
        val index = tags.indexOfFirst { it.id == id }
        if (index >= 0) tags[index] = tags[index].copy(colour = colour)
    }

    fun remove(id: String) {
        tags.removeAll { it.id == id }
        items.remove(id)
    }

    fun assign(nodeId: String, tagIds: Collection<String>) {
        if (tagIds.size > 16) throw AuthFailure("An item can carry at most 16 tags.")
        for (tag in tags) {
            val nodes = items.getOrPut(tag.id) { mutableListOf() }
            nodes.remove(nodeId)
            if (tag.id in tagIds) nodes.add(nodeId)
            if (nodes.isEmpty()) items.remove(tag.id)
        }
    }

    private fun checkName(name: String): String {
        val clean = name.trim().replace(Regex("\\s+"), " ")
        if (clean.isEmpty()) throw AuthFailure("Enter a tag name.")
        if (clean.codePointCount(0, clean.length) > 40) throw AuthFailure("Use a tag name of at most 40 characters.")
        return clean
    }
}

private const val TAGS_KIND = "tags"

/* The registry as the workspace holds it, with the version the next write is based on. */
fun Vault.tags(): Pair<TagRegistry, Int> {
    val ws = workspaceId
    // Kept sealed in the mirror, so the registry reads without network too.
    val key = "tags:$ws"
    val document = try {
        api.document(ws, TAGS_KIND).also { fetched ->
            mirror.putDocument(key, org.json.JSONObject().put("envelope", fetched?.first ?: org.json.JSONObject.NULL).put("version", fetched?.second ?: 0).toString())
        }
    } catch (error: Unreachable) {
        val cached = mirror.document(key)?.let { org.json.JSONObject(it) } ?: throw error
        if (cached.isNull("envelope")) null else cached.getString("envelope") to cached.getInt("version")
    } ?: return TagRegistry.empty() to 0
    val json = documentOpen(DocumentContext(ws, TAGS_KIND, document.second.toULong()), workspaceKey(), base64urlDecode(document.first))
    return TagRegistry.parse(json) to document.second
}

/* Seals and writes the registry one version up; a stale base is reloaded and the change reapplied once. */
fun Vault.saveTags(change: (TagRegistry) -> Unit): TagRegistry {
    val ws = workspaceId
    var attempt = 0
    while (true) {
        val (registry, version) = tags()
        change(registry)
        val envelope = documentSeal(DocumentContext(ws, TAGS_KIND, (version + 1).toULong()), workspaceKey(), registry.toJson())
        try {
            api.putDocument(ws, TAGS_KIND, version, base64urlEncode(envelope))
            return registry
        } catch (error: ApiError) {
            if (error.status != 409 || attempt > 0) throw error
            attempt++
        }
    }
}
